### PR TITLE
search: rewrite lucky job for dynamic query generation

### PR DESCRIPTION
**Pull request by rvantonder**
I rewrote it. The diff is kinda impenetrable. Better to do a live walkthrough review but putting it up for visibility now. Highlights:

- There is now a function to create a Generator, which when called yields the next job to run and the continuation.
- Jobs take the generator, so now it's not possible to enumerate generated functions statically by printing the job (at least not, in the general case where generated queries are unbounded). So: test are updated to create a Generator and just print the generated queries.
- The generated jobs return errs/proposed queries directly related to the query that ran, so those are collected and propagated.

There is room to improve the way generation is structured, I'm not super happy with the control flow, and the job is kinda large now that it got merge with the `Sequential` logic, but it's good progress.

## Test plan
Updated tests.

### CODE REVIEW

File: internal/search/job/jobutil/feeling_lucky_search_job.go [hunk 1]
```diff
@@ -2,66 +2,178 @@ package jobutil
...
+// next is the continuation for the query generator.
+type next func() (job.Job, next)
+
+// NewGenerator creates a new generator using rules and a seed Basic query. It
+// returns a `next` function which, when called, returns the next job
+// generated, and the `next` continuation. A `nil` continuation means query
+// generation is exhausted. Currently it implements a simple strategy that tries
+// to apply rule transforms in order, and generates jobs for transforms that
+// apply successfully.
+func NewGenerator(inputs *run.SearchInputs, seed query.Basic, rules []rule) next {
```
Comment (rvantonder) [inline internal/search/job/jobutil/feeling_lucky_search_job.go:125]: Can be written with object with state tracking + `Yield` method but I didn't have appetite for it and I'm more accustomed to writing this in a pure style. Up for grabs if someone does have appetite, I don't have strong feelings.

File: internal/search/job/jobutil/feeling_lucky_search_job.go [hunk 2]
```diff
@@ -2,66 +2,178 @@ package jobutil
...
+	maxAlerter.Add(alert)
+
+	generated := &alertobserver.ErrLuckyQueries{ProposedQueries: []*search.ProposedQuery{}}
+	var j job.Job
+	for _, next := range f.generators {
+		for {
+			j, next = next()
+			if j == nil && next == nil {
+				break
+			}
```
Comment (camdencheek) [inline internal/search/job/jobutil/feeling_lucky_search_job.go:79]: I think this condition is already covered by the `if j == nil { continue }` on the next line and the `if next == nil { break }` on the last line of the loop. This check seems unneeded and is confusing me

```suggestion
```

File: internal/search/job/jobutil/feeling_lucky_search_job.go [hunk 3]
```diff
@@ -2,66 +2,178 @@ package jobutil
...
+			if ctx.Err() != nil {
+				// Cancellation or Deadline hit implies it's time to stop running jobs.
+				return maxAlerter.Alert, errs
+			}
+
+			var lErr *alertobserver.ErrLuckyQueries
+			if errors.As(err, &lErr) {
+				generated.ProposedQueries = append(generated.ProposedQueries, lErr.ProposedQueries...)
+				errs = errors.Append(errs, generated)
+			}
```
Comment (camdencheek) [inline internal/search/job/jobutil/feeling_lucky_search_job.go:94]: Would this be less awkward if `next()` returned `job, basicQuery, next`? That way we don't have to communicate the generated queries through errors.

Comment (camdencheek) [reply internal/search/job/jobutil/feeling_lucky_search_job.go:94]: Additionally, it looks like we add `generated` to `errs` for every generated query we get, and we also append the error to `errs` on line 97. Won't this send `generated` and the individual `ErrLuckyQueries` many times?

File: internal/search/job/jobutil/feeling_lucky_search_job.go (hunk 1 above)
Comment (camdencheek) [reply internal/search/job/jobutil/feeling_lucky_search_job.go:125]: It's got style

**Review by camdencheek — APPROVED** [review]
Nice, this feels much better. Thanks for following up 🙂

File: internal/search/job/jobutil/feeling_lucky_search_job.go (hunk 2 above)
Comment (rvantonder) [reply internal/search/job/jobutil/feeling_lucky_search_job.go]: The trouble is that, it is possible for `j == nil` and `next == nil` here, when we've exhausted the generator (base case). If we don't break on the base case, we'll exercise

```go
if j == nil {
  continue
}
```

then go around the loop, and call `next()` <= uh oh, nil deref.

This is what I meant I'm not happy with the control flow, but we can make this better later maybe. I've nested the check in attempt to make this clearer, but I don't know that it helps that much.

File: internal/search/job/jobutil/feeling_lucky_search_job.go (hunk 3 above)
Comment (rvantonder) [reply internal/search/job/jobutil/feeling_lucky_search_job.go]: nice catch! I've fixed this up. I didn't notice bad things when testing, didn't try enough things I suppose!

Comment (rvantonder) [reply internal/search/job/jobutil/feeling_lucky_search_job.go]: > Would this be less awkward if next() returned job, basicQuery, next?

Didn't do this part--we need access to rule info too (so `basicQuery` isn't enough) and the multiple return values make this a bit awkward where we call `next`. I Will ponder it though.

**Pull request merged**